modules/integrations/arrow-integrations-retrofit-adapter/src/main/kotlin/arrow.integrations.retrofit.adapter/ResponseCallback.kt

```diff
@@ -0,0 +1,28 @@
+package arrow.integrations.retrofit.adapter
+
+import arrow.core.Either
+import arrow.core.left
+import arrow.core.right
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.HttpException
+import retrofit2.Response
+
+class ResponseCallback<R>(private val proc: (Either<Throwable, R>) -> Unit) : Callback<R> {
+  override fun onResponse(call: Call<R>, response: Response<R>) {
+    if (response.isSuccessful) {
+      val body = response.body()
+      if (body != null) {
+        proc(body.right())
+      } else {
+        proc(IllegalStateException("The request returned a null body").left())
```

Comment:
I've been thinking about it. Doesn't POST and HEAD request imply the possibility of an empty body? Or does retrofit return `null` for failures to deserialize? Maybe it should be `R?` or `Option<R>` instead.

Reply:
Post would be an IO<Unit> or CallK<Unit>. It wouldn't return `null`, it would return Unit. Error to deserialize go straight to `onFailure`. But I should include tests for those cases